pageserver/src/page_service.rs

```diff
@@ -932,8 +933,32 @@ impl postgres_backend_async::Handler for PageServerHandler {
     }
 }
 
-fn get_local_timeline(tenant_id: TenantId, timeline_id: TimelineId) -> Result<Arc<Timeline>> {
-    tenant_mgr::get_tenant(tenant_id, true)
+/// Get active tenant.
+///
+/// If the tenant is Loading, waits for it to become Active, for up to 30 s. That
+/// ensures that queries don't fail immediately after pageserver startup, because
+/// all tenants are still loading.
+async fn get_active_tenant_with_timeout(tenant_id: TenantId) -> Result<Arc<Tenant>> {
+    match tokio::time::timeout(
```

Comment:
Will this actually retry, though? As far as I can tell, we're just reading from the tenant map.

Reply:
[facepalm]. you're right, this is utterly broken.

I pushed a new, significantly rewritten version that should actually do what it says. I also added a test case, with a failpoint to show it.